subprojects/logging/src/main/java/org/gradle/internal/deprecation/DeprecationMessage.java

```diff
@@ -0,0 +1,363 @@
+import org.gradle.api.internal.DocumentationRegistry;
+import org.gradle.internal.featurelifecycle.DeprecatedFeatureUsage;
+
+import java.util.List;
+
+import static org.gradle.internal.deprecation.Messages.methodHasBeenDeprecated;
+import static org.gradle.internal.deprecation.Messages.pleaseUseThisMethodInstead;
+import static org.gradle.internal.deprecation.Messages.pluginHasBeenDeprecated;
+import static org.gradle.internal.deprecation.Messages.propertyHasBeenDeprecated;
+import static org.gradle.internal.deprecation.Messages.thisBehaviourHasBeenDeprecatedAndIsScheduledToBeRemoved;
+import static org.gradle.internal.deprecation.Messages.thisHasBeenDeprecatedAndIsScheduledToBeRemoved;
+import static org.gradle.internal.deprecation.Messages.thisIsScheduledToBeRemoved;
+import static org.gradle.internal.deprecation.Messages.thisWillBecomeAnError;
+import static org.gradle.internal.deprecation.Messages.usingMethodHasBeenDeprecated;
+import static org.gradle.internal.deprecation.Messages.xHasBeenDeprecated;
+
+public class DeprecationMessage {
+
+    private static final DocumentationRegistry DOCUMENTATION_REGISTRY = new DocumentationRegistry();
+
+    private final String summary;
+    private final String removalDetails;
+    private final String advice;
+    private final String contextualAdvice;
+    private final String documentationReference;
+    private final DeprecatedFeatureUsage.Type usageType;
+
+    public DeprecationMessage(String summary, String removalDetails, String advice, String contextualAdvice, String documentationReference, DeprecatedFeatureUsage.Type usageType) {
+        this.summary = summary;
+        this.removalDetails = removalDetails;
+        this.advice = advice;
+        this.contextualAdvice = contextualAdvice;
+        this.documentationReference = documentationReference;
+        this.usageType = usageType;
+    }
+
+    public static class Builder {
+        private String summary;
+        private String removalDetails;
+        private String advice;
+        private String contextualAdvice;
+        private String documentationReference;
+
+        private DeprecatedFeatureUsage.Type usageType = DeprecatedFeatureUsage.Type.USER_CODE_DIRECT;
+
+        public Builder withSummary(String summary) {
+            this.summary = summary;
+            return this;
+        }
+
+        public Builder withRemovalDetails(String removalDetails) {
+            this.removalDetails = removalDetails;
+            return this;
+        }
+
+        public Builder withAdvice(String advice) {
+            this.advice = advice;
+            return this;
+        }
+
+        public Builder withContextualAdvice(String contextualAdvice) {
+            this.contextualAdvice = contextualAdvice;
+            return this;
+        }
+
+        public Builder withDocumentationReference(String documentationReference) {
+            this.documentationReference = documentationReference;
+            return this;
+        }
```

Comment:
This is more of a placeholder now - I will figure the exact type in the next steps.